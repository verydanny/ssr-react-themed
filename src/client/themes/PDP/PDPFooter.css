@value ( RentBlue, Gray1 ) from 'ui/shared/themes/colors/colors.css';

.PDPFooter {
  align-items: center;
  background-color: #fff;
  position: fixed;
  touch-action: none;
  width: calc(100vw - 40px);
  padding: 20px;
  bottom: -100%;
  transition: bottom 1s, visibility 0s linear 2s;
  box-shadow: 0px 0px 10px 4px rgba(0,0,0,0.05);
  visibility: hidden;
  z-index: 10;
}

/* Attribute is added to body element when user scrolls down */
[data-show-pdp-footer] .PDPFooter {
  bottom: 0;
  visibility: visible;
  transition: bottom 1s, visibility 0s linear 0s;
}

.PDPFooter_Info {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 20px;
}

.PDPFooter_InfoContact {
  font-size: 20px;
  color: Gray1;
}

.PDPFooter_Price {
  color: #000;
  font-weight: 600;
  margin: 0 20px;
  align-self: center;
  flex: 0 1 auto;
  font-size: 24px;
}

.PDPFooter .RatingBar {
  width: auto;
  display: flex;
  align-items: center;
}

@value PDPFOOTER_RATING_STAR_WIDTH 20px;
@value PDPFOOTER_RATING_TOTAL_WIDTH calc(5 * PDPFOOTER_RATING_STAR_WIDTH);
.PDPFooter .RatingBar_Background, .PDPFooter .RatingBar_Icons {
  background-size: PDPFOOTER_RATING_STAR_WIDTH;
  height: PDPFOOTER_RATING_STAR_WIDTH;
  width: PDPFOOTER_RATING_TOTAL_WIDTH;
  max-width: PDPFOOTER_RATING_TOTAL_WIDTH;
}

.PDPFooter .RatingBar .Label {
  font-size: 14px;
  line-height: 28px;
  margin-left: 12px;
  vertical-align: middle;
}

.PDPFooter_CTA {
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
}

.PDPFooter_CTA > * {
  background-color: RentBlue;
  border-radius: 10px;
  color: #fff;
  display:block;
  flex-basis: 50%;
  flex-grow: 1;
  font-size: 18px;
  font-weight: 600;
  line-height: 100%;
  margin-left: 20px;
  padding: 16px 0;
  text-align: center;
  text-decoration: none;
  white-space: nowrap;
}

.PDPFooter_CTA > *:first-child {
  margin-left: 0;
}
