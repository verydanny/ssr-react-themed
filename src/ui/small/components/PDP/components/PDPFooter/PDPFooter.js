import React, { PureComponent } from 'react'
import autobind from 'autobind-decorator'
import themed from 'react-themed'
import PropTypes from 'prop-types'
import throttle from 'lodash/throttle'
import AggregateRating from 'ui/shared/components/AggregateRating'

const EVENT = 'scroll'

// Attribute to be added to document.body when scroll percentage is exceded.
// This attribute can be used to show or hide the footer using CSS only.
// Since this attribute is added outside of React it will not cause a re=render.
const ACTIVE_ATTRIBUTE = 'data-show-pdp-footer'

// Percentage of the viewport height where the footer should appear.
const SCROLL_PERCENTAGE = 25

// Avoid computing too much as the scroll event fires
const SCROLL_THROTTLE = 100

@themed(/^PDPFooter/, { pure: true })
class PDPFooter extends PureComponent {
  static propTypes = {
    listing: PropTypes.object.isRequired,
    phoneFormatted: PropTypes.string,
    phoneTel: PropTypes.string,
    price: PropTypes.string,
    theme: PropTypes.object,
    openLeadFormModal: PropTypes.func,
  }

  static defaultProps = {
    theme: {},
  }

  constructor(props) {
    super(props)
    this.handleScroll = throttle(this.handleScroll, SCROLL_THROTTLE)
  }

  componentDidMount() {
    window.addEventListener(EVENT, this.handleScroll)
  }

  componentWillUnmount() {
    window.removeEventListener(EVENT, this.handleScroll)
  }

  @autobind
  handleScroll() {
    const scrollY = window.scrollY
    const viewportHeight = Math.max(document.documentElement.clientHeight, window.innerHeight || 0)
    const scrollPercentage = Math.round((scrollY / viewportHeight) * 100)
    const shouldShowFooter = scrollPercentage > SCROLL_PERCENTAGE

    // Add an attribute to the BODY element when the footer should be showing.
    // The attribute can be used in the CSS rules to show or hide the footer.
    // For example:
    //   .PDPFooter { /* hide */ }
    //   [data-show-pdp-footer] .PDPFooter { /* show */ }
    document.body[shouldShowFooter ? 'setAttribute' : 'removeAttribute'](ACTIVE_ATTRIBUTE, '')
  }

  @autobind
  handleEmailClick() {
    const { listing, openLeadFormModal } = this.props

    if (openLeadFormModal && listing) {
      openLeadFormModal(listing)
    }
  }

  hasPrice() {
    const { price } = this.props
    return Boolean(price)
  }

  hasRatings() {
    const { listing } = this.props
    return Boolean(listing.numRatings)
  }

  renderPrice() {
    const { price, theme } = this.props

    if (!price) return null

    return (
      <span
        key="price"
        data-tid="pdpFooter_price"
        className={theme.PDPFooter_Price}
      >
        {price}
      </span>
    )
  }

  renderRating() {
    const { listing } = this.props
    return (
      <AggregateRating
        key="rating"
        data-tid="pdpFooter_rating"
        listing={listing}
        label={`(${listing.numRatings})`}
      />
    )
  }

  renderInfo() {
    const { theme } = this.props

    const hasInfo = this.hasPrice() || this.hasRatings()

    if (!hasInfo) {
      return (
        <div
          data-tid="pdpFooter_infoContact"
          className={theme.PDPFooter_InfoContact}
        >
          Contact for Price
        </div>
      )
    }

    return [
      this.renderPrice(),
      this.renderRating(),
    ]
  }

  renderPhone() {
    const { phoneTel, phoneFormatted, theme } = this.props

    if (!phoneTel) return null

    // TODO: verify tagging requirements?
    // 'data-listhub_id': this.props.listHubId,
    // 'data-tag_action': 'lead_submission',
    // 'data-tag_item': null,
    // 'data-tag_revenue': this.props.revenue,
    // 'data-tag_selection': 'phone',

    return (
      <a
        data-tid="pdpFooter_phone"
        href={`tel://${phoneTel}`}
        className={theme.PDPFooter_Phone}
      >
        {phoneFormatted}
      </a>
    )
  }

  renderEmail() {
    const { theme } = this.props

    // TODO: verify tagging requirements?
    // data-tag_item="check_availability_button"

    return (
      <button
        data-tid="pdpFooter_contact"
        className={theme.PDPFooter_Contact}
        onClick={this.handleEmailClick}
      >
        Email Property
      </button>
    )
  }

  render() {
    const { theme } = this.props

    return (
      <footer data-tid="pdpFooter" className={theme.PDPFooter}>
        <div className={theme.PDPFooter_Info}>
          {this.renderInfo()}
        </div>
        <div className={theme.PDPFooter_CTA}>
          {this.renderPhone()}
          {this.renderEmail()}
        </div>
      </footer>
    )
  }
}

export default PDPFooter
