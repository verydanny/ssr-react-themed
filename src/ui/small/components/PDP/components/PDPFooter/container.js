import { connect } from 'react-redux'
import pick from 'lodash/fp/pick'
import {
  getListing,
  getPhoneConditionally,
  getPhoneFormatted,
  getPriceTextOnlyPrice,
} from 'app/selectors/pdpListing'
import { openModal } from 'app/store/shared/actions/modal'
import {
  LEAD_MODAL_ID,
} from 'ui/shared/config/modal-const'
import {
  LEAD_MODAL_PROPS,
} from 'ui/small/components/LeadModal/const'
import PDPFooter from './PDPFooter'

const mapStateToProps = state => ({
  listing: getListing(state),
  phoneFormatted: getPhoneFormatted(state),
  phoneTel: getPhoneConditionally(state),
  price: getPriceTextOnlyPrice(state), // priceTextSelector(getListing(state))
})

const mapDispatchToProps = dispatch => ({
  openLeadFormModal: listing => {
    dispatch(openModal(LEAD_MODAL_ID, pick(LEAD_MODAL_PROPS)(listing)))
  },
})

export default connect(mapStateToProps, mapDispatchToProps)(PDPFooter)
