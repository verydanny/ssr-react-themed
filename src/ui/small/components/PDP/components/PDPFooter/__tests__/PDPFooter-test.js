import React from 'react'
import { mount } from 'enzyme'
import { keyMirror } from '@rentpath/react-ui-utils'
import { ThemeProvider } from 'react-themed'
import PDPFooter from '../PDPFooter'

const theme = keyMirror([
  'PDPFooter',
  'PDPFooter_Info',
  'PDPFooter_InfoContact',
  'PDPFooter_Price',
  'PDPFooter_CTA',
  'RatingBar',
  'RatingBar_Background',
  'RatingBar_Icons',
  'Label',
])

const setup = props => mount(
  <ThemeProvider theme={theme}>
    <PDPFooter {...props} />
  </ThemeProvider>
)

describe('small/PDPFooter', () => {
  const props = {
    price: '$1,350',
    phoneTel: '1404678770',
    phoneFormatted: '(404) 678-1234',
    listing: {
      numRatings: 11,
      avgOverallRating: 3.7,
    },
  }

  describe('complete footer', () => {
    const wrapper = setup(props)
    it('contains price, rating, phone link, email button', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('missing price', () => {
    const wrapper = setup({ ...props, price: null })
    it('does not contain the price', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('missing rating', () => {
    const wrapper = setup({ ...props, listing: {} })
    it('does not contain a rating', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('missing price and rating', () => {
    const wrapper = setup({ ...props, price: null, listing: {} })
    it('contains "Contact for Price" and does not contain price or rating', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('missing phone', () => {
    const wrapper = setup({ ...props, phoneTel: null })
    it('does not contain phone link', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('click contact button', () => {
    const callback = jest.fn()
    const wrapper = setup({ ...props, openLeadFormModal: callback })
    wrapper.find('.PDPFooter_CTA button').simulate('click')
    it('calls the openLeadFormModal callback with the listing data', () => {
      expect(callback).toBeCalledWith(
        expect.objectContaining(props.listing)
      )
    })
  })
})
