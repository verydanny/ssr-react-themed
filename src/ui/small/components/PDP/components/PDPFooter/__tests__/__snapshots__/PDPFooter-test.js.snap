// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`small/PDPFooter complete footer contains price, rating, phone link, email button 1`] = `

<footer data-tid="pdpFooter"
        class="PDPFooter"
>
  <div class="PDPFooter_Info">
    <span data-tid="pdpFooter_price"
          class="PDPFooter_Price"
    >
      $1,350
    </span>
    <div class="RatingBar"
         data-tid="pdpFooter_rating"
    >
      <div class="RatingBar_Background">
        <div class="RatingBar_Icons"
             style="width: 70%;"
        >
        </div>
      </div>
      <div class="Label">
        (11)
      </div>
    </div>
  </div>
  <div class="PDPFooter_CTA">
    <a data-tid="pdpFooter_phone"
       href="tel://1404678770"
    >
      (404) 678-1234
    </a>
    <button data-tid="pdpFooter_contact">
      Email Property
    </button>
  </div>
</footer>

`;

exports[`small/PDPFooter missing phone does not contain phone link 1`] = `

<footer data-tid="pdpFooter"
        class="PDPFooter"
>
  <div class="PDPFooter_Info">
    <span data-tid="pdpFooter_price"
          class="PDPFooter_Price"
    >
      $1,350
    </span>
    <div class="RatingBar"
         data-tid="pdpFooter_rating"
    >
      <div class="RatingBar_Background">
        <div class="RatingBar_Icons"
             style="width: 70%;"
        >
        </div>
      </div>
      <div class="Label">
        (11)
      </div>
    </div>
  </div>
  <div class="PDPFooter_CTA">
    <button data-tid="pdpFooter_contact">
      Email Property
    </button>
  </div>
</footer>

`;

exports[`small/PDPFooter missing price and rating contains "Contact for Price" and does not contain price or rating 1`] = `

<footer data-tid="pdpFooter"
        class="PDPFooter"
>
  <div class="PDPFooter_Info">
    <div data-tid="pdpFooter_infoContact"
         class="PDPFooter_InfoContact"
    >
      Contact for Price
    </div>
  </div>
  <div class="PDPFooter_CTA">
    <a data-tid="pdpFooter_phone"
       href="tel://1404678770"
    >
      (404) 678-1234
    </a>
    <button data-tid="pdpFooter_contact">
      Email Property
    </button>
  </div>
</footer>

`;

exports[`small/PDPFooter missing price does not contain the price 1`] = `

<footer data-tid="pdpFooter"
        class="PDPFooter"
>
  <div class="PDPFooter_Info">
    <div class="RatingBar"
         data-tid="pdpFooter_rating"
    >
      <div class="RatingBar_Background">
        <div class="RatingBar_Icons"
             style="width: 70%;"
        >
        </div>
      </div>
      <div class="Label">
        (11)
      </div>
    </div>
  </div>
  <div class="PDPFooter_CTA">
    <a data-tid="pdpFooter_phone"
       href="tel://1404678770"
    >
      (404) 678-1234
    </a>
    <button data-tid="pdpFooter_contact">
      Email Property
    </button>
  </div>
</footer>

`;

exports[`small/PDPFooter missing rating does not contain a rating 1`] = `

<footer data-tid="pdpFooter"
        class="PDPFooter"
>
  <div class="PDPFooter_Info">
    <span data-tid="pdpFooter_price"
          class="PDPFooter_Price"
    >
      $1,350
    </span>
  </div>
  <div class="PDPFooter_CTA">
    <a data-tid="pdpFooter_phone"
       href="tel://1404678770"
    >
      (404) 678-1234
    </a>
    <button data-tid="pdpFooter_contact">
      Email Property
    </button>
  </div>
</footer>

`;
